crates/wadm-types/src/validation.rs

```diff
@@ -520,31 +522,41 @@ fn check_dangling_links(manifest: &Manifest) -> Vec<ValidationFailure> {
     failures
 }
 
-fn check_secrets_mapped_to_policies(manifest: &Manifest) -> Vec<ValidationFailure> {
+/// Ensure that a manifest has secrets that are mapped to known policies
+/// and that those policies have the expected type and properties.
+fn validate_secrets_policies(manifest: &Manifest) -> Vec<ValidationFailure> {
     let policies = manifest.policy_lookup();
     let mut failures = Vec::new();
     for c in manifest.components() {
         for secret in c.secrets() {
-            if !policies.contains_key(&secret.source.policy) {
-                failures.push(ValidationFailure::new(
+            match policies.get(&secret.properties.policy) {
+                Some(policy) => {
+                    if !policy.properties.contains_key("backend") {
+                        failures.push(ValidationFailure::new(
+                            ValidationFailureLevel::Error,
+                            format!(
+                                "secret '{}' is mapped to policy '{}' which does not include a 'backend' property",
+                                secret.name, secret.properties.policy
+                            ),
+                        ))
+                    }
+                    if policy.policy_type != SECRET_POLICY_TYPE {
+                        failures.push(ValidationFailure::new(
+                            ValidationFailureLevel::Error,
+                            format!(
+                                "policy '{}' type did not match required type '{}'",
+                                policy.name, SECRET_POLICY_TYPE
+                            ),
+                        ));
+                    }
+                }
```

Comment:
A few things:
* we will likely have other policies that wadm doesn't know about, so we shouldn't have a validation failure if the policy type isn't a secret policy
* The `backend` key should only be checked if policy.policy_type != SECRET_POLICY_TYPE`

Reply:
Great catch. I thought this was only called from the secret context, let me fix this.